motoko/encrypted-notes-dapp/src/encrypted_notes_rust/src/lib.rs

```diff
@@ -0,0 +1,591 @@
+use candid::CandidType;
+use ic_cdk::api::{caller as caller_api};
+use ic_cdk::export::{candid, Principal};
+use ic_cdk::storage;
+use ic_cdk_macros::*;
+use serde::{Deserialize, Serialize};
+use std::cell::RefCell;
+use std::collections::BTreeMap;
+use std::mem;
+use std::collections::btree_map::Entry::{Vacant, Occupied};
+
+type PrincipalName = String;
+type PublicKey = String;
+type Ciphertext = String;
+type DeviceAlias = String;
+
+#[derive(Clone, CandidType, Deserialize)]
+pub enum GetCiphertextError {
+    // Ensure API types in [encrypted_notes_rust.did] are named exactly as specified below
+    #[serde(rename = "notFound")]
+    NotFound,
+    #[serde(rename = "notSynced")]
+    NotSynced,
+}
+
+#[derive(Clone, CandidType, Deserialize)]
+pub enum Result {
+    // Ensure API types in [encrypted_notes_rust.did] are named exactly as specified below
+    #[serde(rename = "err")]
+    Err(GetCiphertextError),
+    #[serde(rename = "ok")]
+    Ok(Ciphertext),
+}
+
+/// Deriving CandidType or implementing it is necessary for
+/// almost everything IC - if you want your structs to
+/// Save in stable storage or serialize in inputs/outputs
+/// You should derive CandidType, Serialize, Deserialize.
+#[derive(Clone, CandidType, Serialize, Deserialize)]
+pub struct EncryptedNote {
+    id: u128,
+    encrypted_text: String,
+}
+
+/// There can only be one Type in stable storage at a time.
+/// We use this struct to represent the full CanisterState
+/// So we can serialize it to stable storage.
+#[derive(Clone, CandidType, Serialize, Deserialize)]
+struct CanisterState {
+    counter: u128,
+    // We use a BTreeMap vice a HashMap for deterministic ordering.
+    notes: BTreeMap<PrincipalName, Vec<EncryptedNote>>,
+    user_store: BTreeMap<Principal, UserStore>,
+}
+
+#[derive(Clone, CandidType, Serialize, Deserialize, Default)]
+pub struct UserStore {
+    // We use a BTreeMap vice a HashMap for deterministic ordering.
+    device_list: BTreeMap<DeviceAlias, PublicKey>,
+    ciphertext_list: BTreeMap<PublicKey, Ciphertext>,
+}
+
+impl UserStore {
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+// WASM is single-threaded by nature. [RefCell] and [thread_local!] is used vice unsafe.
+// This is to ensure that the canister state can be used throughout.
+// Your other option here is to avoid [thread_local!] and use a [RefCell<RwLock/Mutex/Atomic>].
+// Here we use [thread_local!] because it is simpler.
+thread_local! {
+
+    // Define dapp limits - important for security assurance
```

Comment:
My back of the envelope calculation gives 1GB of notes, is this what you had in mind?

Reply:
My calculations are:
1. Each user may store ~1 GB worth of notes
2. Each user may store up to ~2 MB for device management (negligible)

Since we permit up to 1000 users, our constraints do not ensure that canister limit of 4 GB is respected.

Reply:
Shouldn't we limit the number of notes to a smaller value to ensure the canister does not run out of memory?
E.g. 5000 notes per user still seems plenty :)

Reply:
Addressed in https://github.com/dfinity/examples/pull/164/commits/dceba856d20ed24916b34a33819968896946557b